packages/embark-transaction-manager/src/lib/eth_accounts.ts

```diff
@@ -0,0 +1,61 @@
+import { Callback, Embark, Events, Logger } /* supplied by @types/embark in packages/embark-typings */ from "embark";
+import Web3 from "web3";
+const { blockchain: blockchainConstants } = require("embark-core/constants");
+import { __ } from "embark-i18n";
+import { AccountParser, dappPath } from "embark-utils";
+import { EventEmitter } from "events";
+import Transaction from "./Transaction";
+
+const TXS_TO_MODIFY = [
+  blockchainConstants.transactionMethods.eth_accounts,
+  blockchainConstants.transactionMethods.personal_listAccounts,
+];
+
+function arrayEqual(arrayA: string[], arrayB: string[]) {
+  if (arrayA.length !== arrayB.length) {
+    return false;
+  } else {
+    return arrayA.every((address, index) => Web3.utils.toChecksumAddress(address) === Web3.utils.toChecksumAddress(arrayB[index]));
+  }
+}
+
+export default class EthAccounts implements Transaction {
+  public readonly events: Events;
+  public readonly logger: Logger;
+  constructor(readonly embark: Embark, readonly web3: Web3, public nodeAccounts: any[], public accounts: any[], readonly txEvents: EventEmitter) {
+    this.events = embark.events;
+    this.logger = embark.logger;
+
+    if (accounts.length) {
+      this.embark.registerActionForEvent("blockchain:proxy:response", this.checkResponseFor_eth_accounts.bind(this));
+    }
+  }
+
+  private async checkResponseFor_eth_accounts(params: any, callback: Callback<any>) {
+
+    if (!(TXS_TO_MODIFY.includes(params.reqData.method))) {
+      return callback(null, params);
+    }
+
+    this.logger.trace(__(`Modifying blockchain '${params.reqData.method}' response:`));
+    this.logger.trace(__(`Original request/response data: ${JSON.stringify(params)}`));
+
+    try {
+      let accounts = this.accounts;
+      const nodeAccounts = params.respData.result;
+      if (!arrayEqual(nodeAccounts, this.nodeAccounts)) {
+        // reset backing variables so accounts is recalculated
+        this.txEvents.emit("nodeAccounts:updated", params.respData.result);
+        // update accounts for this response only.
+        accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, this.web3, dappPath(), this.logger, nodeAccounts);
```

Comment:
Those two lines could be replaced by a request instead of an emit, because then, the tx-manager would have time to modify the accounts, then this flow can continue and `this.accounts` are the new accounts (so no need to parse twice)

Reply:
Great idea. Updated.